### PR TITLE
Added key-derivation JavaScript examples analog to python version

### PR DESCRIPTION
**Pull request by AlexanderBuzz**
I've added the [Cryptographic Key Derivation](https://github.com/XRPLBounties/Proposals/blob/main/bounties/0%20JavaScript%20code%20samples.md) Milestone 1 from the JavaScript code samples bounty proposal. It's inspired by the existing python version and delivers the same output.

### CODE REVIEW

**Comment by AlexanderBuzz** [comment]
Output analog to python version, e.g.:

Seed (base58): sp5DoqZTVsmnvFzuuvgU8kt7xyW2a
Seed (hex): 0123848AD5018C7304A7FD4D022CE3CE
Seed (rippled RFC-1751): REST HAW AD HUED SHIM FLED DINT BULL DOOR FUM TRY FIRE
            
Ed25519 Secret Key (hex): ED245D5F9F257814A5928F5B8EFBC6E93932DB1515EC0FEEBB1B9C78D74A201A1F
Ed25519 Public Key (hex): ED05BE431B72AA5C462056A91BD5690AE0CF0CC8DC72B278F5FFD6AD8B353C1EA0
Ed25519 Public Key (base58 - Account): aKNoywKWp5bvfLvS7WgUKUXvr6rKKg9RUS7sRbyR8MJbCHimtDy5
            
secp256k1 Secret Key (hex): 00947FA3B1B5C4249B68F2724024B82A8DFA7C050BE564804AD852B8C168DF6B8B
secp256k1 Public Key (hex): 02921E3C08C4C8BBD9ED0A3E5F71C5AACC3FF9003F6A3F43F19842574AC2A686CC
secp256k1 Public Key (base58 - Account): aBPhiPkBYhYgwuYUqMAs3iFfNX22oRnqV5Hriauib1gGhBzfK362
secp256k1 Public Key (base58 - Validator): n9KeiGDtJz35MHSFxqw8Ze96VVBrxVBkMcvavvVUwXQHn1bLZpNi

File: content/_code-samples/key-derivation/js/index.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+        //from hex formatted seed
+        if (isHex(seedValue)) {
+            const decoded = hexToBytes(seedValue);
+            if(decoded.length === 16) {
+                return decoded
+            } else {
+                //Raise Error
+                console.log('Raise error!')
```
Comment (jonathanlei) [inline content/_code-samples/key-derivation/js/index.js]:
```suggestion
                throw new Error("incorrect decoded seed length")
```

**Review by jonathanlei — APPROVED** [review]

File: content/_code-samples/key-derivation/js/index.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+                return key;
+            }
+        }
+        // This error is practically impossible to reach.
+        // The order of the curve describes the (finite) amount of points on the curve
+        // https://github.com/indutny/elliptic/blob/master/lib/elliptic/curves.js#L182
+        // How often will an (essentially) random number generated by Sha512 be larger than that?
+        // There's 2^32 chances (the for loop) to get a number smaller than the order,
+        // and it's rare that you'll even get past the first loop iteration.
+        // Note that in TypeScript we actually need the throw, otherwise the function signature would be BN | undefined
```
Comment (mDuo13) [inline content/_code-samples/key-derivation/js/index.js:144]: :100:

Comment (AlexanderBuzz) [reply content/_code-samples/key-derivation/js/index.js:144]: Well, when toying around with this for the PHP-XRPL implementation, I wondered why this potentially infinite loop is needed at all when in most cases i is 0 or 1. Managed to get to 43 in a buggy in-between state of the code :)

**Review by AlexanderBuzz — COMMENTED** [review]
Seems OK, but the link checker action needs to run again

**Comment by JST5000** [comment]
For reference, this is the updated link to the bounty this is relating to: https://github.com/XRPLBounties/Proposals/blob/main/bounties/0032%20Python%20Code%20Samples.md

**Comment by AlexanderBuzz** [comment]
> For reference, this is the updated link to the bounty this is relating to: https://github.com/XRPLBounties/Proposals/blob/main/bounties/0032%20Python%20Code%20Samples.md

Hi, I guess this is the correct one, as this is about the JavaScript examples: https://github.com/XRPLBounties/Proposals/blob/main/bounties/0031%20JavaScript%20Code%20Samples.md

Anyways, I will now claim that Bounty, perhaps someone will trigger that link checker pipeline someday :)

**Comment by JST5000** [comment]
> Hi, I guess this is the correct one, as this is about the JavaScript examples: https://github.com/XRPLBounties/Proposals/blob/main/bounties/0031%20JavaScript%20Code%20Samples.md

Ah sorry for posting the wrong link!

File: content/_code-samples/key-derivation/js/index.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+const Secp256k1 = elliptic.ec('secp256k1');
+const { rfc1751MnemonicToKey, keyToRFC1751Mnemonic } = require("xrpl/dist/npm/Wallet/rfc1751");
+const hashjs = require("hash.js");
+const Sha512 = require("ripple-keypairs/dist/Sha512")
+const { codec, encodeAccountPublic, encodeNodePublic } = require("ripple-address-codec");
+
+const XRPL_SEED_PREFIX = 0x21
+
+const isHex = function(value) {
+    const regex = new RegExp(/^0x([0-9a-f]+(\.[0-9a-f]*)?|\.[0-9a-f]+)(p[+-]?\d+)?$/, 'i')
```
Comment (JST5000) [inline content/_code-samples/key-derivation/js/index.js]: Why does this regex have `\.` in it? 
Do you have examples of inputs which would trigger the logic beyond the first `/^0x[0-9a-f]+`?

Also, I think this might not hit upper case hex, which should be considered hex

**Comment by mDuo13** [comment]
The link check error is benign—the only failure is a known issue that was in the master branch for a little while and has since been fixed by #1522.

File: content/_code-samples/key-derivation/js/index.js (hunk 3 above)
Comment (mDuo13) [reply content/_code-samples/key-derivation/js/index.js]: The `,'i'` at the end makes this RegExp case-insensitive, so that part is OK.

I'm not sure what `(\.[0-9a-f]*)?|\.[0-9a-f]+)(p[+-]?\d+)??` is about though. This regex accepts `0x1234.1` as an example value, which is not really correct.

For what it's worth, in XRP Ledger code, we usually don't use a prefix like `0x` for hex strings either.

It's a drop-in replacement given how this function is used elsewhere, but I would prefer this code use a simpler regex like the following:

```suggestion
    const regex = new RegExp(/^[0-9a-f]+$/, 'i')
```

(You could also constrain the length, for example by replacing the `+` with `{2,64}` which results in 1-32 bytes when decoded since two hex characters makes 1 byte.)

File: content/_code-samples/key-derivation/js/index.js [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    
+    constructor(seedValue = '', checkRFC1751 = false) {
+        this.bytes = this._initSeed(seedValue)
+        this.ED25519Keypair = this._deriveED25519Keypair(this.bytes)
+        this.Secp256K1Keypair = this._deriveSecp256K1Keypair(this.bytes)
+    }
+
+    _initSeed(seedValue) {
+
+        //default to random seed
```
Comment (JST5000) [inline content/_code-samples/key-derivation/js/index.js]: Style note - please add a space and capitalize the first letter in comments
```suggestion
        // Default to random seed
```

File: content/_code-samples/key-derivation/js/index.js [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            // We hash the bytes to find a 256 bit number, looping until we are sure it
+            // is less than the order of the curve.
+            const hasher = new Sha512.default().add(bytes);
+            // If the optional discriminator index was passed in, update the hash.
+            if (discrim !== undefined) {
+                hasher.addU32(discrim);
+            }
+            hasher.addU32(i);
+            const key = hasher.first256BN();
+            /* istanbul ignore else */
```
Comment (JST5000) [inline content/_code-samples/key-derivation/js/index.js:132]: Why is this required?

File: content/_code-samples/key-derivation/js/index.js [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+if (process.argv[2] !== undefined) {
+    console.log(process.argv[2]);
+    const seed = new Seed(process.argv[2]);
+    console.log(seed.toString())
+} else {
+    const seed = new Seed();
+    console.log(seed.toString())
+}
+
```
Comment (JST5000) [inline content/_code-samples/key-derivation/js/index.js]: Probably worth adding a section here explaining why the commented out code exists, ex. "// To experiment with this code, uncomment the examples below:"

File: content/_code-samples/key-derivation/js/index.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+//from hex formatted seed
+//const hexSeed = "0x21edc3dec3ef1873cf8f333381c5f360c3"
+//const seed = new Seed(hexSeed);
+//console.log(seed.toString());
+
+//from rippled RFC1751 mnemonic
+//const rippledMnemonic = 'KANT FISH GENE COST WEB JAKE CHUM HAD SOD MID KICK BOTH'
+//const seed = new Seed(rippledMnemonic);
+//console.log(seed.toString());
```
Comment (JST5000) [inline content/_code-samples/key-derivation/js/index.js]: I would remove the fromMnemonic part of this implementation as it's a deprecated feature of rippled to generate these mnemonics, it's only been added to the library to have backwards compatibility with new accounts generated using the old `wallet_propose` method, but we're planning on deprecating the client library support for it.

File: content/_code-samples/key-derivation/js/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+    npm start "sEdSKaCy2JT7JaM7v95H9SxkhP9wS2r"
+
+### Rippled RFC1751 Mnemonic
+
+    npm start "KANT FISH GENE COST WEB JAKE CHUM HAD SOD MID KICK BOTH"
+
+### Hex formatted seed
+
+    npm start "0x21edc3dec3ef1873cf8f333381c5f360c3"
```
Comment (JST5000) [inline content/_code-samples/key-derivation/js/README.md]: Currently when I run this the snippet crashes: 

```
    if (!val) throw new Error(msg || 'Assertion failed');
              ^

Error: Invalid character in 0x21edc3dec3ef1873cf8f333381c5f360c3
    at assert (/Users/jacksonmills/Documents/node_modules/bn.js/lib/bn.js:6:21)
    at parseHex4Bits (/Users/jacksonmills/Documents/node_modules/bn.js/lib/bn.js:198:7)
    at parseHexByte (/Users/jacksonmills/Documents/node_modules/bn.js/lib/bn.js:203:13)
    at BN._parseHex (/Users/jacksonmills/Documents/node_modules/bn.js/lib/bn.js:225:13)
    at BN.init [as _init] (/Users/jacksonmills/Documents/node_modules/bn.js/lib/bn.js:103:14)
    at new BN (/Users/jacksonmills/Documents/node_modules/bn.js/lib/bn.js:39:12)
    at hexToBytes (/Users/jacksonmills/Documents/xrpl-dev-portal-bounty-alex/xrpl-dev-portal/test.js:31:36)
    at Seed._initSeed (/Users/jacksonmills/Documents/xrpl-dev-portal-bounty-alex/xrpl-dev-portal/test.js:60:29)
    at new Seed (/Users/jacksonmills/Documents/xrpl-dev-portal-bounty-alex/xrpl-dev-portal/test.js:45:27)
    at Object.<anonymous> (/Users/jacksonmills/Documents/xrpl-dev-portal-bounty-alex/xrpl-dev-portal/test.js:195:18)
```

**Review by JST5000 — CHANGES_REQUESTED** [review]
I'm currently unable to run the formatted seed version of this code snippet without running into an error. My guess would be that is because of the regex not recognizing `x` in the given string, but I could be wrong.

File: content/_code-samples/key-derivation/js/README.md (hunk 8 above)
Comment (AlexanderBuzz) [reply content/_code-samples/key-derivation/js/README.md]: Fixed, I removed the "0x" and now it works flawlessly. Probably a copy-paste error from PHP code, been switching back and forth from PHP to JavaScript.

File: content/_code-samples/key-derivation/js/index.js (hunk 4 above)
Comment (AlexanderBuzz) [reply content/_code-samples/key-derivation/js/index.js]: I've updated all comments according to styleguide

File: content/_code-samples/key-derivation/js/index.js (hunk 6 above)
Comment (AlexanderBuzz) [reply content/_code-samples/key-derivation/js/index.js]: I have removed that section, it is not useful anymore

File: content/_code-samples/key-derivation/js/index.js (hunk 7 above)
Comment (AlexanderBuzz) [reply content/_code-samples/key-derivation/js/index.js]: It is gone, that this is deprecated somehow slipped me

File: content/_code-samples/key-derivation/js/index.js (hunk 5 above)
Comment (AlexanderBuzz) [reply content/_code-samples/key-derivation/js/index.js:132]: This function is taken directly from xrpl/ripple-keypairs, including the funny comments...

**Comment by AlexanderBuzz** [comment]
> I'm currently unable to run the formatted seed version of this code snippet without running into an error. My guess would be that is because of the regex not recognizing `x` in the given string, but I could be wrong.

This is fixed now

**Comment by AlexanderBuzz** [comment]
Eh, wrong button...

**Comment by JST5000** [comment]
> Eh, wrong button...

Hahaha, it happens. I'll check this out now

**Review by JST5000 — APPROVED** [review]
LGTM

**Review by mDuo13 — APPROVED** [review]

**Comment by JST5000** [comment]
@AlexanderBuzz please submit to claim this bounty here: [https://xrpl.typeform.com/xrpl-bounties](https://xrpl.typeform.com/xrpl-bounties?utm_source=xxxxx&utm_medium=xxxxx&utm_campaign=xxxxx&utm_term=xxxxx&utm_content=xxxxx)
